### ISSUE TITLE
MultiHeadAttention quietly ignores masking when annotated with tf.function

### ISSUE THREAD

**Issue by tsdeng** [keras-team/keras#17655]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): MacOS 13.2.1
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.10
- Python version: 3.10
- Bazel version (if compiling from source): NA
- GPU model and memory: Apple Silicone M2 Max
- Exact command to reproduce:

Run following script in a Jupyter notebook. The outputs of with_mask and without_mask are the same. Mask is not working. When `@tf.function` is used.

```
import tensorflow as tf
class ApplyMHA(tf.keras.layers.Layer):
    def __init__(self, **kwargs):
        super().__init__()
        self.mha = tf.keras.layers.MultiHeadAttention(**kwargs)

    @tf.function
    def call(self, x):
        return self.mha(query=x, key=x, value=x)
x = tf.convert_to_tensor(
    [[1,2,0,0],
     [2,3,1,0]
    ]
)

# Create to embedding tables, one with masking, one without.
# They use the same seed as initializer so the two tables should be identical except masking.
initializer = tf.keras.initializers.RandomUniform(
    minval=-0.05, maxval=0.05, seed=123)

embedding_table_with_mask = tf.keras.layers.Embedding(input_dim=100, output_dim=3, mask_zero=True, embeddings_initializer=)
embedding_table_without_mask = tf.keras.layers.Embedding(input_dim=100, output_dim=3, mask_zero=False, embeddings_initializer=tf.keras.initializers.RandomUniform(
    minval=-0.05, maxval=0.05, seed=123
))
embedding_with_mask = embedding_table_with_mask(x)
print("embedding_with_mask:", embedding_with_mask)
embedding_without_mask = embedding_table_without_mask(x)
print("embedding_without_mask:", embedding_without_mask)

mha = ApplyMHA(num_heads=2, key_dim=3)
print("===After applying MHA====")

# The outputs of with_mask and without_mask are the same. Mask is not working.
print("with_mask:", mha(embedding_with_mask))
print("without_mask:", mha(embedding_without_mask))
```

**Describe the problem**.

In the script provided above, when MultiHeadAttention is called from a function annotated with `@tf.function`, it quietly ignores the masks from the input tensor. This is demonstrated by the results from "with_mask" and "without_mask" being the same.

Notice this problem only happens when calling from a context that's using `@tf.function`. Removing `@tf.function` will makes is issue disappear.

IIUC the root cause of this is MultiHeadAttention(MHA) relies on `_keras_mask` attached to the input tensor([code](https://github.com/keras-team/keras/blob/e6784e4302c7b8cd116b74a784f4b78d60e83c26/keras/layers/attention/multi_head_attention.py#L642)). `_keras_mask` is not available in the context of `@tf.function`. To fix the problem, MHA should rely on the `mask` variable passed to the `call` function, but this will be a pretty big change.

**Describe the current behavior**.
When MultiHeadAttention is called from a function annotated with `@tf.function`, it quietly ignores the masks from the input tensor. 

**Describe the expected behavior**.
MultiHeadAttention should respect the masking from inputs when being called from a function annotated with `@tf.function`

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no):
yes.
Change the call signature of MHA to accept an array of inputs and their masks. Do not rely on `tensor._keras_mask`

**Standalone code to reproduce the issue**.

Mentioned in `Exact command to reproduce` section.

**Source code / logs**.

Mentioned in `Exact command to reproduce` section.

**Comment by SuryanarayanaY** [comment]
Hi @tsdeng ,

Thanks for reporting this. I have replicated the reported behaviour on MHA masking w.r.t `tf.function`.Please refer to attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/9ef96ee90918cbc4bd6d81ecf7ca4cc4/17655-mha-masking-using-tf_function.ipynb).

But I am not sure on the root cause that you mentioned.I just tried the code,
 `query_mask = getattr(query, "_keras_mask", None)` with and without `tf.function` decorator and this is returning same result `None` for both cases. The same also attached in above gist.

Could you please check and share your thoughts ?

**Comment by tsdeng** [comment]
Hi @SuryanarayanaY I think there is an error in your code. You were calling your function with `x`, instead you should pass in `embedding_with_mask`.

The issue can be demonstrated [here](https://colab.research.google.com/drive/1VPip5MtoHBtlSCO2MwhfnrI9PcDBHw06#scrollTo=-zyBNjHmqpUV&line=5&uniqifier=1). It shows when tf.function is not used, the mask can be accessed successfully.

**Comment by SuryanarayanaY** [comment]
I have replicated the issue and attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/08d5dfe842c81a150ee5c1b9827f8463/17655-mha-masking-using-tf_function-r1.ipynb). It seems the problem is valid as masking ignored quietly in `MultiHeadAttention` when annotated with `tf.function`.

@tsdeng , If you willing to raise PR please feel to do. Thanks!
